### PR TITLE
Gluster Volume & Brick Utilization Alert

### PR DESCRIPTION
**Pull request by umangachapagain**
Fixes : gluster/gluster-mixins#4

Signed-off-by: Umanga Chapagain <chapagainumanga@gmail.com>

### CODE REVIEW

File: gluster-utilization.libsonnet [hunk 1]
```diff
@@ -0,0 +1,66 @@
+{
+  prometheusAlerts+:: {
+    groups+: [
+      {
+        name: 'gluster-utilization',
+        rules: [
+          {
+            alert: 'GlusterVolumeUtilization',
+            expr: |||
+                  100 * gluster:volume_capacity_used_bytes_total:sum > 80,
```
Comment (shtripat) [inline gluster-utilization.libsonnet]: Isnt this metric `gluster:volume_capacity_used_bytes_total` no of bytes? If so how is percentage getting calculated here?

File: gluster-utilization.libsonnet [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            |||,
+            'for': '5m',
+            labels: {
+              severity: 'warning',
+            },
+            annotations: {
+              message: 'Gluster Volume Utilization more than 80%',
```
Comment (shtripat) [inline gluster-utilization.libsonnet]: Alert message could be `Gluster volume {volname} crossed 80% usage`. Same in other places as well.

File: gluster-utilization.libsonnet [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+              severity: 'critical',
+            },
+            annotations: {
+              message: 'Gluster Volume Utilization more than 90%',
+            },
+          },
+          {
+            alert: 'GlusterBrickUtilization',
+            expr: |||
+                  100 * gluster_brick_capacity_used_bytes
```
Comment (shtripat) [inline gluster-utilization.libsonnet]: Something similar is required for volume capacity alerts as well (division by total no of bytes)

File: gluster-utilization.libsonnet [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    groups+: [
+      {
+        name: 'gluster-utilization',
+        rules: [
+          {
+            alert: 'GlusterVolumeUtilization',
+            expr: |||
+                  100 * gluster:volume_capacity_used_bytes_total:sum > 80,
+            |||,
+            'for': '5m',
```
Comment (shtripat) [inline gluster-utilization.libsonnet]: How to make this time `5m` configurable?

File: gluster-utilization.libsonnet (hunk 3 above)
Comment (umangachapagain) [reply gluster-utilization.libsonnet]: ack.

File: gluster-utilization.libsonnet (hunk 2 above)
Comment (cloudbehl) [reply gluster-utilization.libsonnet]: @umangachapagain any update on this?

Comment (umangachapagain) [reply gluster-utilization.libsonnet]: fixed.

File: gluster-utilization.libsonnet (hunk 4 above)
Comment (umangachapagain) [reply gluster-utilization.libsonnet]: @shtripat  @cloudbehl  made it configurable.

File: alerts/gluster-utilization.libsonnet [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            expr: |||
+                  100 * gluster_brick_capacity_used_bytes{%(glusterExporterSelector)s}
+                      / gluster_brick_capacity_bytes_total{%(glusterExporterSelector)s} > 90
+            ||| % $._config,
+            'for': $._config.volumeUtilizationAlertTime,
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              message: 'Gluster Brick {{$s.brick}} Utilization more than 90%',
```
Comment (cloudbehl) [inline alerts/gluster-utilization.libsonnet]: labels it will be.

**Comment by shtripat** [comment]
> I don't see any label({{$labels.brick}}) name as `brick` in https://github.com/gluster/gluster-prometheus/blob/master/docs/metrics.adoc#gluster_brick_capacity_used_bytes.
> 
> Is there something else you are using?

I feel we rather should use `brick_path` label here. It would make more sense it we combine `host` and `brick_path` in the alert message which would clearly mention which brick on which node is problematic.

**Comment by cloudbehl** [comment]
@shtripat correct, my intention was to get the hostname+path only. As we are doing in tendrl

@umangachapagain please do the required changes.

File: alerts/gluster-utilization.libsonnet (hunk 5 above)
Comment (umangachapagain) [reply alerts/gluster-utilization.libsonnet]: fixed in subsequent PR.

**Comment by umangachapagain** [comment]
> @shtripat correct, my intention was to get the hostname+path only. As we are doing in tendrl
> 
> @umangachapagain please do the required changes.

Done.

**Comment by cloudbehl** [comment]
@shtripat @anmolsachan can you review it once and also it will be good if someone can test it. 

@umangachapagain also can you paste some test result you did and steps on how to test it. So will be easy for everyone to verify.

**Comment by cloudbehl** [comment]
Looks good to me also.

@umangachapagain verified?

**Comment by cloudbehl** [comment]
@umangachapagain can you squash the commits

**Comment by umangachapagain** [comment]
alerts-list
brick alert details

Note: Setting threshold at 5% for testing purpose only.

**Comment by anmolsachan** [comment]
The PR looks good, But shouldn't we have GlusterBrickUtilization and GlusterVolumeUtilization as recording rules and put thresholds on that? This will be a calculative operation and its better to do it via recording rules and use the O/P.

File: alerts/gluster-utilization.libsonnet [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+              severity: 'warning',
+            },
+            annotations: {
+              message: 'Gluster Brick {{$labels.host}}:{{$labels.brick_path}} Utilization more than 80%',
+            },
+          },
+          {
+            alert: 'GlusterBrickUtilization',
+            expr: |||
+              100 * gluster_brick_capacity_used_bytes{%(glusterExporterSelector)s}
```
[the remark this answers is not in the document]
Comment (umangachapagain) [reply alerts/gluster-utilization.libsonnet]: why do we need a recording rule when we already have required metrics?

1. We get fresh metrics in configured intervals.
2. We match our expressions against these metrics in configured intervals.
3. The expression has to be true for a mentioned time duration which is configurable.
4. Then we generate the alert.

[the remark this answers is not in the document]
Comment (cloudbehl) [reply alerts/gluster-utilization.libsonnet]: @anmolsachan We can think about it.

But the only concern that we need to make for warning then critical. for all the metric I am not in favor of that. @shtripat @anmolsachan thoughts?

Also, we have already recording rule for the volume utilization. https://github.com/gluster/gluster-mixins/pull/14

[the remark this answers is not in the document]
Comment (shtripat) [reply alerts/gluster-utilization.libsonnet]: @umangachapagain @cloudbehl I agree. We should write recording rules only for required cases. Lets not just keep creating metrics. If existing metrics could be used for generation of alerts based on calculations, that's good enough.

[the remark this answers is not in the document]
Comment (umangachapagain) [reply alerts/gluster-utilization.libsonnet]: > But the only concern that we need to make for warning then critical. for all the metric I am not in favor of that. @shtripat @anmolsachan thoughts?
> 

These were in the proposal and discussed. 
However, we can always customize things if needed, like using predict_linear instead or percentage.

[the remark this answers is not in the document]
Comment (anmolsachan) [reply alerts/gluster-utilization.libsonnet]: @anmolsachan I understand the points. This PR should be good for now.

**Comment by umangachapagain** [comment]
Depends on gluster/gluster-mixins/pull/17 . @shtripat

**Comment by shtripat** [comment]
@umangachapagain rebase and resolve conflicts

**Pull request merged**